File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/AlchemyCategory.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.math.Point;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.Renderer;
import me.shedaniel.rei.api.client.gui.widgets.Widget;
import me.shedaniel.rei.api.client.gui.widgets.Widgets;
import me.shedaniel.rei.api.client.registry.display.DisplayCategory;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.GlyphWidget;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;

import java.util.ArrayList;
import java.util.List;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public class AlchemyCategory implements DisplayCategory<AlchemyDisplay> {

    public static final int ARROW_U = 0;
    public static final int ARROW_V = 0;
    public static final Identifier GLYPHS = id("textures/gui/rei/glyphs.png");
    public static final int MARGIN = 6;
    public static final int HEIGHT = 3 * 18 + MARGIN * 2;
    public static final int BARRELS_Y = HEIGHT / 2;
    public static final int ARROW_Y = BARRELS_Y - 18;
    public static final int REACTANT_Y = ARROW_Y;
    public static final int PLUS_Y = REACTANT_Y;
    public static final int CATALYST_Y = PLUS_Y;
    public static final int PRODUCT_Y = ARROW_Y;
    public static final int BYPRODUCT_Y = PRODUCT_Y;
    public static final int SPAWN_Y = BYPRODUCT_Y;
    public static final int PLUS_U = 3 * 16;
    public static final int PLUS_V = 0;
    public static final int WIDTH = 8 * 18 + MARGIN * 2;
    public static final int BARRELS_X = WIDTH / 2 - 9;
    public static final int ARROW_X = BARRELS_X;
    public static final int REACTANT_X = ARROW_X - 18;
    public static final int PLUS_X = REACTANT_X - 18;
    public static final int CATALYST_X = PLUS_X - 18;
    public static final int PRODUCT_X = ARROW_X + 18;
    public static final int BYPRODUCT_X = PRODUCT_X + 18;
    public static final int SPAWN_X = BYPRODUCT_X + 18;
    private final ItemStack icon;
    private final String name;

    public AlchemyCategory(ItemStack icon, String name) {
        this.icon = icon;
        this.name = name;
    }

    @Override
    public CategoryIdentifier<? extends AlchemyDisplay> getCategoryIdentifier() {
        return PluginEntry.ALCHEMY;
    }

    @Override
    public int getDisplayHeight() {
        return HEIGHT;
    }

    @Override
    public int getDisplayWidth(AlchemyDisplay display) {
        return WIDTH;
    }

    @Override
    public Renderer getIcon() {
        return EntryStacks.of(icon);
    }

    @Override
    public Text getTitle() {
        return new TranslatableText(this.name);
    }

    @Override
    public List<Widget> setupDisplay(AlchemyDisplay display, Rectangle bounds) {
        var widgets = new ArrayList<Widget>();
        widgets.add(Widgets.createRecipeBase(bounds));

        // Plus Glyph
        widgets.add(new GlyphWidget(bounds, bounds.getMinX() + PLUS_X, bounds.getMinY() + PLUS_Y, 16, 16, GLYPHS, PLUS_U, PLUS_V));
        // Arrow Glyph
        widgets.add(new GlyphWidget(bounds, bounds.getMinX() + ARROW_X, bounds.getMinY() + ARROW_Y, 16, 16, GLYPHS, ARROW_U, ARROW_V));

        var outputs = display.getOutputEntries();
        var catalyst = display.getCatalyst().get(0);
        var reactant = display.getReactant().get(0);

        var barrels = display.getBarrel();
        var product = outputs.get(0);
        var byproduct = outputs.get(1);
        var toSpawn = outputs.get(2);

        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + REACTANT_X, bounds.getMinY() + REACTANT_Y)).entries(reactant));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + CATALYST_X, bounds.getMinY() + CATALYST_Y)).entries(catalyst));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + PRODUCT_X, bounds.getMinY() + PRODUCT_Y)).entries(product));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + BYPRODUCT_X, bounds.getMinY() + BYPRODUCT_Y)).entries(byproduct));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + SPAWN_X, bounds.getMinY() + SPAWN_Y)).entries(toSpawn));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + BARRELS_X, bounds.getMinY() + BARRELS_Y)).entries(barrels));

        return widgets;
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/AlchemyCategory.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.compatibility.rei.GlyphWidget
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.id
import exnihilofabrico.util.asREIEntry
import exnihilofabrico.util.getExNihiloItemStack
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.api.RecipeCategory
import me.shedaniel.rei.gui.widget.EntryWidget
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import me.shedaniel.rei.gui.widget.Widget
import java.util.function.Supplier

class AlchemyCategory: RecipeCategory<AlchemyDisplay> {

    override fun getIdentifier() = PluginEntry.ALCHEMY
    override fun getLogo() = getExNihiloItemStack("oak_barrel").asREIEntry()
    override fun getCategoryName() = "Alchemy"


    override fun getDisplayHeight() = HEIGHT
    override fun getDisplayWidth(display: AlchemyDisplay) =
        WIDTH

    override fun setupDisplay(displaySupplier: Supplier<AlchemyDisplay>, bounds: Rectangle): MutableList<Widget> {
        val display = displaySupplier.get()
        val widgets = mutableListOf<Widget>(RecipeBaseWidget(bounds))

        val plusGlyph = GlyphWidget(bounds, bounds.minX + PLUS_X, bounds.minY + PLUS_Y, GLYPH_WIDTH, GLYPH_HEIGHT, GLYPHS, PLUS_U, PLUS_V)
        val arrowGlyph = GlyphWidget(bounds, bounds.minX + ARROW_X, bounds.minY + ARROW_Y, GLYPH_WIDTH, GLYPH_HEIGHT, GLYPHS, ARROW_U, ARROW_V)
        widgets.add(plusGlyph)
        widgets.add(arrowGlyph)

        val reactant = display.inputEntries[0]
        val catalyst = display.inputEntries[1]
        val barrels = display.inputEntries[2]
        val product = display.outputEntries[0]
        val byproduct = display.outputEntries[1]
        val toSpawn = display.outputEntries[2]

        widgets.add(EntryWidget.create(bounds.minX + REACTANT_X, bounds.minY + REACTANT_Y).entries(reactant))
        widgets.add(EntryWidget.create(bounds.minX + CATALYST_X, bounds.minY + CATALYST_Y).entries(catalyst))
        widgets.add(EntryWidget.create(bounds.minX + PRODUCT_X, bounds.minY + PRODUCT_Y).entry(product))
        widgets.add(EntryWidget.create(bounds.minX + BYPRODUCT_X, bounds.minY + BYPRODUCT_Y).entry(byproduct))
        widgets.add(EntryWidget.create(bounds.minX + SPAWN_X, bounds.minY + SPAWN_Y).entry(toSpawn))
        widgets.add(EntryWidget.create(bounds.minX + BARRELS_X, bounds.minY + BARRELS_Y).entries(barrels))

        return widgets
    }

    companion object {
        val GLYPHS = id("textures/gui/rei/glyphs.png")

        val MARGIN = 6

        val WIDTH = 8*18 + MARGIN*2
        val HEIGHT = 3*18 + MARGIN*2

        val BARRELS_X = WIDTH/2 - 9
        val BARRELS_Y = HEIGHT/2 + 9

        val ARROW_X = BARRELS_X
        val ARROW_Y = BARRELS_Y - 18

        val REACTANT_X = ARROW_X - 18
        val REACTANT_Y = ARROW_Y

        val PLUS_X = REACTANT_X - 18
        val PLUS_Y = REACTANT_Y

        val CATALYST_X = PLUS_X - 18
        val CATALYST_Y = PLUS_Y
        val PRODUCT_X = ARROW_X + 18
        val PRODUCT_Y = ARROW_Y
        val BYPRODUCT_X = PRODUCT_X + 18
        val BYPRODUCT_Y = PRODUCT_Y
        val SPAWN_X = BYPRODUCT_X + 18
        val SPAWN_Y = BYPRODUCT_Y

        val GLYPH_WIDTH = 16
        val GLYPH_HEIGHT = 16
        val ARROW_U = 0
        val ARROW_V= 0

        val PLUS_U = 3*16
        val PLUS_V= 0
    }

}